cilium_eip_no_masquerade_agent/src/main.rs

```diff
@@ -98,36 +81,125 @@ impl k8s_controller::Context for Context {
+                        }),
+                        PatchOperation::Remove(RemoveOperation {
+                            path: finalizer_path,
+                        }),
+                    ])),
+                )
+                .await
+                .unwrap();
+        }
+    }
+
+    async fn wait_for_chain_to_exist(&self) -> Result<(), Box<dyn std::error::Error>> {
+        info!("Waiting for {CHAIN} chain to exist");
+        while !self.iptables.chain_exists(TABLE, CHAIN)? {
+            trace!("{CHAIN} does not exist, sleeping");
+            sleep(Duration::from_secs(1)).await;
+        }
+        Ok(())
+    }
+
+    async fn get_eni_indexes_for_existing_ip_rules(
+        &self,
+    ) -> Result<HashSet<u32>, rtnetlink::Error> {
+        let rules = self.ip_rule_handle.get(IpVersion::V4).execute();
+        rules
+            .try_filter(|rule| future::ready(rule.nlas.contains(&Nla::FwMask(0xf))))
+            .map_ok(|rule| {
+                rule.nlas
+                    .into_iter()
+                    .find_map(|nla| match nla {
+                        Nla::FwMark(i) => Some(i),
+                        _ => None,
+                    })
+                    .expect("There should always be an associated mark if there is a mask.")
+            })
+            .collect::<Vec<Result<u32, rtnetlink::Error>>>()
+            .await
+            .into_iter()
+            .collect()
+    }
+
+    fn ensure_iptables_rule(&self, rule: &str) -> Result<(), Box<dyn std::error::Error>> {
+        if !self.iptables.exists(TABLE, CHAIN, rule)? {
+            event!(Level::INFO, rule = ?rule, "Appending iptables rule");
+            self.iptables.append(TABLE, CHAIN, rule)?;
+        }
+        Ok(())
+    }
+
+    fn ensure_restore_mark_iptables_rule(&self) -> Result<(), Box<dyn std::error::Error>> {
+        let rule = "-i lxc+ -m comment --comment \"cilium: secondary interfaces\" -j CONNMARK --restore-mark --nfmask 0xf --ctmask 0xf";
+        self.ensure_iptables_rule(rule)?;
+        Ok(())
+    }
+
+    fn ensure_mark_iptables_rule(
+        &self,
+        interface_index: u32,
+    ) -> Result<(), Box<dyn std::error::Error>> {
+        let rule = format!("-i eth{interface_index} -m comment --comment \"cilium: eth{interface_index}\" -m addrtype --dst-type UNICAST --limit-iface-in -j CONNMARK --set-xmark 0x{interface_index}/0xf");
+        self.ensure_iptables_rule(&rule)?;
+        Ok(())
+    }
+
+    async fn insert_ip_rule(&self, interface_index: u32) -> Result<(), rtnetlink::Error> {
+        info!("Inserting ip rule for eth{interface_index}");
+        // ip rule add pref 100 from all fwmark 0x{i}/0xf lookup {10 + i}
+        let mut rule_add_request = self
+            .ip_rule_handle
+            .add()
+            .v4()
+            .table_id(10 + interface_index)
+            .action(1) // Not sure what this action is, but it seems to work
+            .priority(100);
```

Comment:
any context on this number?

Reply:
This is somewhat arbitrary, but it is before the rules Cilium injects for the eth0 marks and for in-VPC traffic, but after the local and pod inbound rules. We also previously used 100 in the older version.

```
0:      from all lookup local
20:     from all to 10.2.43.28 lookup main
20:     from all to 10.2.40.118 lookup main
20:     from all to 10.2.46.85 lookup main
100:    from all fwmark 0x1/0xf lookup 11
100:    from all fwmark 0x2/0xf lookup 12
100:    from all fwmark 0x3/0xf lookup 13
100:    from all fwmark 0x4/0xf lookup 14
100:    from all fwmark 0x5/0xf lookup 15
100:    from all fwmark 0x6/0xf lookup 16
100:    from all fwmark 0x7/0xf lookup 17
100:    from all fwmark 0x8/0xf lookup 18
100:    from all fwmark 0x9/0xf lookup 19
100:    from all fwmark 0xa/0xf lookup 20
100:    from all fwmark 0xb/0xf lookup 21
100:    from all fwmark 0xc/0xf lookup 22
100:    from all fwmark 0xd/0xf lookup 23
100:    from all fwmark 0xe/0xf lookup 24
100:    from all fwmark 0xf/0xf lookup 25
109:    from all fwmark 0x80/0x80 lookup main
111:    from 10.2.40.51 to 10.2.0.0/16 lookup 11
111:    from 10.2.43.28 to 10.2.0.0/16 lookup 11
111:    from 10.2.40.118 to 10.2.0.0/16 lookup 11
111:    from 10.2.46.85 to 10.2.0.0/16 lookup 11
32766:  from all lookup main
32767:  from all lookup default
```

Reply:
I've added a comment.